lib/winapi/timezoneapi.c

```diff
@@ -0,0 +1,196 @@
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_DATE, &type, &standardDate, sizeof(standardDate), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_DLT_BIAS, &type, &dayLightBias, sizeof(dayLightBias), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_BIAS, &type, &standardBias, sizeof(standardBias), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_DLT_NAME, &type, dayLightName, sizeof(dayLightName), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+    queryStatus = ExQueryNonVolatileSetting(XC_TZ_STD_NAME, &type, standardName, sizeof(standardName), NULL);
+    if (!NT_SUCCESS(queryStatus)) {
+        SetLastError(RtlNtStatusToDosError(queryStatus));
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    if (standardDate.Month > 12 || dayLightDate.Month > 12 || 
+        standardDate.DayOfWeek > 6 || dayLightDate.DayOfWeek > 6 ||
+        standardDate.Hour > 23 || dayLightDate.Hour > 23) {
+        return TIME_ZONE_ID_INVALID;
+    }
+
+    // Xbox stores up to 4 characters here.
+    for (UCHAR i = 0; i < 4; i++) {
+        lpTimeZoneInformation->StandardName[i] = standardName[i];
+        lpTimeZoneInformation->DaylightName[i] = dayLightName[i];
+    }
+
+    // There was no Daylight savings date info
+    if (standardDate.Month == 0 || dayLightDate.Month == 0) {
+        return TIME_ZONE_ID_UNKNOWN;
+    }
+
+    lpTimeZoneInformation->StandardDate.wMonth = standardDate.Month;
+    lpTimeZoneInformation->StandardDate.wDayOfWeek = standardDate.DayOfWeek;
+    lpTimeZoneInformation->StandardDate.wDay = standardDate.Day;
+    lpTimeZoneInformation->StandardDate.wHour = standardDate.Hour;
+    lpTimeZoneInformation->StandardBias = standardBias;
+
+    lpTimeZoneInformation->DaylightDate.wMonth = dayLightDate.Month;
+    lpTimeZoneInformation->DaylightDate.wDayOfWeek = dayLightDate.DayOfWeek;
+    lpTimeZoneInformation->DaylightDate.wDay = dayLightDate.Day;
+    lpTimeZoneInformation->DaylightDate.wHour = dayLightDate.Hour;
+    lpTimeZoneInformation->DaylightBias = dayLightBias;
+
+    // Now we need to determine if the current time is within the day light savings range
+    KeQuerySystemTime(&kTime);
+    RtlTimeToTimeFields(&kTime, &dateNow);
+
+    // Determine what day in the month DST starts and ends
+    UCHAR startDay = GetActualDay(dateNow.Year, &dayLightDate);
+    UCHAR endDay = GetActualDay(dateNow.Year, &standardDate);
```

Comment:
nit: The variable names for `dayLightDate` and `standardDate` confused me in the section below and also make these names inconsistent (should be `dayLightDay` and `standardDay`).

However, I think having `daylightStartDay` from `daylightStartDate` & `daylightEndDay` from `daylightStartDate` would make this entire function clearer, even if this conflicts with the Xbox EEPROM naming convention.

Reply:
I think ive done this. Its now all daylightStartDate, daylightEndDate, daylightStartDay, daylightEndDay etc.

Reply:
nit: Would have preferred `daylight` over `dayLight` to be consistent with existing Xbox names (`Daylight`). But it's just nitpicking, so you can resolve this conversation (with or without change).

Reply:
Done, all daylight